lib/decoder_impl.cc

```diff
@@ -19,804 +19,1021 @@
+                }
+            }
 
-    unsigned int decoder_impl::get_shift_fft(gr_complex* samples) {
-        float fft_mag[d_number_of_bins];
-        gr_complex mult_hf[d_samples_per_symbol];
+            if (!found_change) {
+                return 0.0f;
+            }
 
-        #ifdef CFO_CORRECT
-            determine_cfo(&samples[0]);
-            d_debug << "CFO: " << d_cfo_estimation << std::endl;
-            correct_cfo(&samples[0], d_samples_per_symbol);
-        #endif
+            max = std::max_element(samples + gr::lora::clamp((int)(max - coeff),  0, (int)window),
+                                   samples + gr::lora::clamp(max +        coeff, 0u,      window)) - samples;
+            min = std::min_element(samples + gr::lora::clamp(max +           1u, 0u,      window),
+                                   samples + gr::lora::clamp(max +   2u * coeff, 0u,      window)) - samples;
 
-        samples_to_file("/tmp/data", &samples[0], d_samples_per_symbol, sizeof(gr_complex));
+            max += (min - max) / 2;
 
-        // Multiply with ideal downchirp
-        for(uint32_t i = 0; i < d_samples_per_symbol; i++) {
-            mult_hf[i] = conj(samples[i] * d_downchirp[i]);
-        }
+            *index = max;
 
-        samples_to_file("/tmp/mult", &mult_hf[0], d_samples_per_symbol, sizeof(gr_complex));
+            // Extra to allow falling edge on 99% of sample == Practically synced
+            // Should improve detection by not discarding almost synced samples, but doesn't..
+//            if ((max - coeff / 2) > (window * 0.98f)) {
+////                printf("ALREADY SYNCED\n");
+//                *index = 0;
+//                return 0.99f;
+//            }
 
-        // Perform decimation
-        for (uint32_t i = 0; i < d_number_of_bins; i++) {
-            firdecim_crcf_execute(d_decim, &mult_hf[d_decim_factor*i], &d_mult[i]);
-        }
+            float max_correlation = 0.0f;
 
-        samples_to_file("/tmp/resampled", &d_mult[0], d_number_of_bins, sizeof(gr_complex));
+            const float average = std::accumulate(samples + max, samples + window, 0.0f) / (window - max);
+            const float au      = std::accumulate(&this->d_upchirp_ifreq[max], &this->d_upchirp_ifreq[window], 0.0f) / (window - max);
+            const float sd      = this->stddev(&samples[max], window - max, average) * this->stddev(&this->d_upchirp_ifreq[max], window - max, au);
 
-        // Perform FFT
-        fft_execute(d_q);
+            for (uint32_t j = max; j < window; j++) {
+                max_correlation += (samples[j] - average) * (this->d_upchirp_ifreq[j] - au) / sd;
+            }
 
-        // Get magnitude
-        for(int i = 0; i < d_number_of_bins; i++) {
-            fft_mag[i] = abs(d_fft[i]);
+            max_correlation /= (float)(window - max);
+            
+            return max_correlation;
+
+            /* NEWER *********************************************************************/
```

Comment:
Instead of commenting experimental implementations of the upchirp detection I would put them in a separate function with a fitting name.